### PR TITLE
devenv: add initial docker development environment

### PR DESCRIPTION
**Pull request by EmbeddedAndroid**
## Description
As per https://github.com/stacks-network/sbtc/issues/85 we desire a standalone, docker based development environment for sBTC. 

## Checklist
- [ ] Additions and modifications are documented
- [ ] Additions and modifications are tested

### CODE REVIEW

File: devenv/stacks-api/docker-compose-stacks-api.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    environment:
+      - NODE_ENV=production
+      - GIT_TAG=master
+      - PG_HOST=postgres
+      - PG_PORT=5432
+      - PG_USER=postgres
+      - PG_PASSWORD=postgres
+      - PG_DATABASE=postgres
+      - STACKS_CHAIN_ID=0x80000000
+      - V2_POX_MIN_AMOUNT_USTX=90000000260
```
[the remark this answers is not in the document]
Comment (friedger) [reply devenv/stacks-api/docker-compose-stacks-api.yml:34]: Also the unlock height for pox-2, pox-3 looks far too big

image

[the remark this answers is not in the document]
Comment (friedger) [reply devenv/stacks-api/docker-compose-stacks-api.yml:34]: When are the hardforks scheduled?

File: devenv/bitcoin/docker-compose-bitcoin.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    container_name: bitcoin
+    build:
+      context: ./docker
+   command: bash -c "/srv/explorer/run.sh bitcoin-regtest explorer"
+   ports:
+     - 18444:18444
+     - 18443:18443
+     - 8094:80
+   environment:
+     - NO_REGTEST_MINING=1
```
Comment (friedger) [inline devenv/bitcoin/docker-compose-bitcoin.yml:15]: Can we mine a few bitcoin blocks at start? 100 like in clarinet integrate?

File: devenv/stacks/docker/Config.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+[node]
+working_dir = "/root/stacks-node/data"
+rpc_bind = "0.0.0.0:20443"
+p2p_bind = "0.0.0.0:20444"
```
Comment (friedger) [inline devenv/stacks/docker/Config.toml:4]: I see a few warnings like this. Maybe we can stop these to increase the performance?
`WARN [1694100874.608488] [src/net/p2p.rs:2484] [ThreadId(24)] local.80000000://(bind=0.0.0.0:20444)(pub=None): failed to learn public IP: NoSuchNeighbor
`

File: devenv/mine.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -x
+
+./bitcoin/bin/bitcoin-cli createwallet test
```
Comment (friedger) [inline devenv/mine.sh:5]: This fails if the wallet already exists.

File: devenv/README.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+cd devenv/sbtc
+./build
+```
+
+Now you can use the sbtc cli by calling
+
+```
+./devenv/sbtc/bin/sbtc <args>
+```
+
```
Comment (friedger) [inline devenv/README.md:109]: How do I launch the Romeo engine?

**Review by friedger — DISMISSED** [review]
This looks good so far.

I left some comments where we need more improvements :-)

File: devenv/sbtc/up.sh [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+ docker compose -f docker-compose-sbtc.yml up -d
```
[the remark this answers is not in the document]
Comment (friedger) [reply devenv/sbtc/up.sh:3]: Actually, sbtc works fine.
This is about romeo engine binary

File: devenv/README.md (hunk 5 above)
Comment (friedger) [reply devenv/README.md:109]: @EmbeddedAndroid And how to launch the bridge UI ?

**Comment by friedger** [comment]
I wonder whether it is easier to run clarinet integrate in a docker Container? Is that possible? @EmbeddedAndroid

**Review by stjepangolemac — COMMENTED** [review]
I think this PR is beautiful. 😍 

If it doesn't take too long to spin everything up I think we can ignore persistence for now. Also, from the comments it seems there's still a couple of things to add so I'm not approving yet but I'm looking forward to approve.

I'll try to play with it when I find time. Thank you!

**Comment by stjepangolemac** [comment]
The only reason we picked Esplora was to have something running quicker — without spinning up another Bitcoin node. @friedger is now working on a PR that will introduce an alternative bitcoin client to Romeo that goes directly through RPC, so you might not even need to have Esplora here.

Long term I also think it's better to rely on RPC directly than Esplora.

When it comes to the bitcoind version I think we've relied on the newest one because the `bdk` crate we use already relies on descriptors. Here's a sample usage:

https://github.com/stacks-network/sbtc/blob/master/sbtc-cli/src/commands/utils.rs#L23-L35

File: devenv/README.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+This is a collection of Dockerized services to provide a simple 
+standalone development environment to sBTC developers. It also 
+includes some helper scripts to make it easier to operate.
+
+## Docker and Docker Compose
+
+To use this environment, you must install:
+
+- [Docker](https://docs.docker.com/engine/install/)
```
Comment (jo-tm) [inline devenv/README.md:11]: Is there any specific Docker version required?

File: devenv/README.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+```
+cd devenv/
+./mine.sh <number of blocks>
+```
+Once you mine blocks on bitcoin, there is a delay before the 
+stacks mines those blocks, and relays the data to the API. If
+your explorer is crashing, just wait a few minutes and try again.
+
+## Services
```
Comment (jo-tm) [inline devenv/README.md:74]: are there any default accounts for the services or everyone is open access without accounts?

File: devenv/log.sh [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+docker logs -f $1
```
Comment (jo-tm) [inline devenv/log.sh:3]: another script for wiping `wipe` or `clean` could be useful for hard reset of the env.

**Review by jo-tm — COMMENTED** [review]
left some comments on README and maybe another command `clean` or `wipe`

**Comment by friedger** [comment]
@stjepangolemac is working on bitcoin node rpc client. Maybe we don't need esplora in the end.

**Comment by friedger** [comment]
@EmbeddedAndroid it should be the responsibility of the binary to deploy the dependency contracts. @stjepangolemac is wishing on it

**Comment by stjepangolemac** [comment]
Yeah it's problematic at the moment because there are multiple contracts to deploy. I'm looking the fix / extension of the deployment phase.

**Comment by stjepangolemac** [comment]
Stacks Explorer can't be built on my Mac. Looking into it.

<details>
```
=> ERROR [explorer build 6/7] RUN pnpm i                                                                                                                                                  9.6s
------                                                                                                                                                                                          
 > [explorer build 6/7] RUN pnpm i:                                                                                                                                                             
0.607 Lockfile is up to date, resolution step is skipped                                                                                                                                        
0.645 Progress: resolved 1, reused 0, downloaded 0, added 0                                                                                                                                     
0.796 Packages: +779                                                                                                                                                                            
0.796 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++                                                                                                          
1.646 Progress: resolved 779, reused 0, downloaded 111, added 111
2.647 Progress: resolved 779, reused 0, downloaded 204, added 204
3.647 Progress: resolved 779, reused 0, downloaded 339, added 339
4.650 Progress: resolved 779, reused 0, downloaded 422, added 421
5.649 Progress: resolved 779, reused 0, downloaded 612, added 612
6.650 Progress: resolved 779, reused 0, downloaded 701, added 701
7.657 Progress: resolved 779, reused 0, downloaded 775, added 774
8.302 Progress: resolved 779, reused 0, downloaded 779, added 779, done
8.555 .../node_modules/bufferutil install$ node-gyp-build
8.555 .../node_modules/utf-8-validate install$ node-gyp-build
8.555 .../iltorb@2.4.5/node_modules/iltorb install$ node ./scripts/install.js || node-gyp rebuild
8.571 .../node_modules/next-plugin-preact postinstall$ node ./postinstall.js
8.610 .../node_modules/@sentry/cli install$ node ./scripts/install.js
8.640 .../node_modules/next-plugin-preact postinstall: [PREACT] Required manual step!
8.640 .../node_modules/next-plugin-preact postinstall: Install the alias packages:
8.641 .../node_modules/next-plugin-preact postinstall:   npm i --save react@npm:@preact/compat react-dom@npm:@preact/compat
8.641 .../node_modules/next-plugin-preact postinstall: or:
8.641 .../node_modules/next-plugin-preact postinstall:   yarn add react@npm:@preact/compat react-dom@npm:@preact/compat
8.648 .../node_modules/next-plugin-preact postinstall: Done
8.691 .../node_modules/@sentry/cli install: [sentry-cli] Downloading from https://downloads.sentry-cdn.com/sentry-cli/1.75.0/sentry-cli-Linux-aarch64
8.697 .../iltorb@2.4.5/node_modules/iltorb install: info looking for cached prebuild @ /root/.npm/_prebuilds/30bebf-iltorb-v2.4.5-node-v93-linuxmusl-arm64.tar.gz
8.697 .../iltorb@2.4.5/node_modules/iltorb install: http request GET https://github.com/nstepien/iltorb/releases/download/v2.4.5/iltorb-v2.4.5-node-v93-linuxmusl-arm64.tar.gz
8.721 .../node_modules/bufferutil install: gyp info it worked if it ends with ok
8.722 .../node_modules/bufferutil install: gyp info using node-gyp@9.4.0
8.723 .../node_modules/bufferutil install: gyp info using node@16.20.2 | linux | arm64
8.733 .../node_modules/utf-8-validate install: gyp info it worked if it ends with ok
8.734 .../node_modules/utf-8-validate install: gyp info using node-gyp@9.4.0
8.734 .../node_modules/utf-8-validate install: gyp info using node@16.20.2 | linux | arm64
8.764 .../node_modules/bufferutil install: gyp ERR! find Python 
8.764 .../node_modules/bufferutil install: gyp ERR! find Python Python is not set from command line or npm configuration
8.764 .../node_modules/bufferutil install: gyp ERR! find Python Python is not set from environment variable PYTHON
8.764 .../node_modules/bufferutil install: gyp ERR! find Python checking if "python3" can be used
8.764 .../node_modules/bufferutil install: gyp ERR! find Python - "python3" is not in PATH or produced an error
8.764 .../node_modules/bufferutil install: gyp ERR! find Python checking if "python" can be used
8.764 .../node_modules/bufferutil install: gyp ERR! find Python - "python" is not in PATH or produced an error
8.764 .../node_modules/bufferutil install: gyp ERR! find Python 
8.764 .../node_modules/bufferutil install: gyp ERR! find Python **********************************************************
8.764 .../node_modules/bufferutil install: gyp ERR! find Python You need to install the latest version of Python.
8.765 .../node_modules/bufferutil install: gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
8.765 .../node_modules/bufferutil install: gyp ERR! find Python you can try one of the following options:
8.765 .../node_modules/bufferutil install: gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
8.765 .../node_modules/bufferutil install: gyp ERR! find Python   (accepted by both node-gyp and npm)
8.765 .../node_modules/bufferutil install: gyp ERR! find Python - Set the environment variable PYTHON
8.765 .../node_modules/bufferutil install: gyp ERR! find Python - Set the npm configuration variable python:
8.765 .../node_modules/bufferutil install: gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
8.765 .../node_modules/bufferutil install: gyp ERR! find Python For more information consult the documentation at:
8.765 .../node_modules/bufferutil install: gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
8.765 .../node_modules/bufferutil install: gyp ERR! find Python **********************************************************
8.765 .../node_modules/bufferutil install: gyp ERR! find Python 
8.765 .../node_modules/bufferutil install: gyp ERR! configure error 
8.765 .../node_modules/bufferutil install: gyp ERR! stack Error: Could not find any Python installation to use
8.765 .../node_modules/bufferutil install: gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:330:47)
8.765 .../node_modules/bufferutil install: gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:159:21)
8.766 .../node_modules/bufferutil install: gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:202:16)
8.766 .../node_modules/bufferutil install: gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:294:16)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at exithandler (node:child_process:410:5)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at ChildProcess.errorhandler (node:child_process:422:5)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at onErrorNT (node:internal/child_process:485:16)
8.767 .../node_modules/bufferutil install: gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
8.767 .../node_modules/bufferutil install: gyp ERR! System Linux 5.15.49-linuxkit-pr
8.767 .../node_modules/bufferutil install: gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
8.767 .../node_modules/bufferutil install: gyp ERR! cwd /app/node_modules/.pnpm/bufferutil@4.0.7/node_modules/bufferutil
8.767 .../node_modules/bufferutil install: gyp ERR! node -v v16.20.2
8.767 .../node_modules/utf-8-validate install: gyp ERR! find Python 
8.767 .../node_modules/utf-8-validate install: gyp ERR! find Python Python is not set from command line or npm configuration
8.767 .../node_modules/utf-8-validate install: gyp ERR! find Python Python is not set from environment variable PYTHON
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python checking if "python3" can be used
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python - "python3" is not in PATH or produced an error
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python checking if "python" can be used
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python - "python" is not in PATH or produced an error
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python 
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python **********************************************************
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python You need to install the latest version of Python.
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python you can try one of the following options:
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python   (accepted by both node-gyp and npm)
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python - Set the environment variable PYTHON
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python - Set the npm configuration variable python:
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python For more information consult the documentation at:
8.768 .../node_modules/utf-8-validate install: gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
8.769 .../node_modules/utf-8-validate install: gyp ERR! find Python **********************************************************
8.769 .../node_modules/utf-8-validate install: gyp ERR! find Python 
8.769 .../node_modules/utf-8-validate install: gyp ERR! configure error 
8.769 .../node_modules/utf-8-validate install: gyp ERR! stack Error: Could not find any Python installation to use
8.769 .../node_modules/utf-8-validate install: gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:330:47)
8.769 .../node_modules/utf-8-validate install: gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:159:21)
8.769 .../node_modules/utf-8-validate install: gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:202:16)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/lib/find-python.js:294:16)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at exithandler (node:child_process:410:5)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at ChildProcess.errorhandler (node:child_process:422:5)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at onErrorNT (node:internal/child_process:485:16)
8.770 .../node_modules/utf-8-validate install: gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
8.770 .../node_modules/utf-8-validate install: gyp ERR! System Linux 5.15.49-linuxkit-pr
8.771 .../node_modules/utf-8-validate install: gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/pnpm/dist/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
8.773 .../node_modules/utf-8-validate install: gyp ERR! cwd /app/node_modules/.pnpm/utf-8-validate@6.0.2/node_modules/utf-8-validate
8.774 .../node_modules/utf-8-validate install: gyp ERR! node -v v16.20.2
8.774 .../node_modules/utf-8-validate install: gyp ERR! node-gyp -v v9.4.0
8.774 .../node_modules/utf-8-validate install: gyp ERR! not ok 
8.774 .../node_modules/bufferutil install: gyp ERR! node-gyp -v v9.4.0
8.774 .../node_modules/bufferutil install: gyp ERR! not ok 
8.775 .../node_modules/utf-8-validate install: Failed
8.777  ELIFECYCLE  Command failed with exit code 1.
8.777 .../node_modules/bufferutil install: Failed
------
failed to solve: process "/bin/sh -c pnpm i" did not complete successfully: exit code: 1
```
</details>

**Comment by stjepangolemac** [comment]
Okay I've spent some time on this and it feels unstable or / and unclear. Here are some things I discovered:

- stacks-explorer couldn't build without adding python3, make, and g++ to its `apk add` command in Dockerfile
- stacks node needs a lot of time to start mining, a lot more than 3 minutes (I think uncommenting `inv_sync_interval` and `download_interval` in it's config maybe speeds it up?)
- when I connect to the devnet from my chrome extension leather wallet it doesn't show the right addresses (cc @markmhx)
  - if I connect to devnet that it auto discovers main btc address is testnet, while ordinal address is regtest (with regtest being correct)

image

  - if I add a new network entry and manually type the details for local connection main btc address is testnet, and ordinal address is also testnet (both wrong)

image

- stacks node premines stx to some addresses without any details on where to find mnemonics and private keys for them (@friedger then told me they seem to be the same as ones hardcoded in clarinet)
- when the stacks node managed to mine something I couldn't figure out how to fund those accounts with btc (I ultimately figured out I can get the right bcrt1 addresses from electrum if I import the key there but  as already mentioned it's not visible in the leather wallet)

**Review by friedger — DISMISSED** [review]
Amazing!
This works for me!
LGTM!

**Review by stjepangolemac — APPROVED** [review]
Given everything we've discovered and mentioned today I think there's a lot of value of having this in `main` at the moment. I invite you to merge this @EmbeddedAndroid and continue fixing in subsequent branches.

**Review by friedger — APPROVED** [review]
LGTM and be so much more productive

**Pull request merged**